packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -0,0 +1,461 @@
+        }));
+
+        emit ExitPoolRequested(msg.sender, poolAmountIn, minAmountsOut);
+    }
+
+    // Anyone is able to shut down the pool when requests aren't being processed any more.
+    function shutdown()
+        external
+        payable
+        online
+    {
+        require(
+            block.timestamp > queue[queuePos].timestamp + MAX_AGE_REQUEST_UNTIL_POOL_SHUTDOWN &&
+            queue[queuePos].status == Status.PENDING,
+            "REQUEST_NOT_TOO_OLD"
+        );
+
+        if (!exchange.isInWithdrawalMode()) {
+            for (uint i = 0; i < tokens.length; i++) {
+                exchange.forceWithdraw{value: msg.value/tokens.length}(
+                    address(this),
+                    tokens[i].addr,
+                    accountID
+                );
+            }
+        }
+
+        shutdownTimestamp = block.timestamp;
+    }
+
+    // Only used to withdraw from the pool when shutdown.
+    // Otherwise LPs should withdraw by doing normal queued exit requests.
+    function withdrawFromPoolWhenShutdown(uint poolAmountIn)
+        external
+        offline
+    {
+        bool ready = true;
+        if (exchange.isInWithdrawalMode()) {
+            // Check if all tokens were withdrawn using Merkle proofs
+            for (uint i = 0; i < tokens.length; i++) {
+                ready = ready && !exchange.isWithdrawnInWithdrawalMode(accountID, tokens[i].addr);
+            }
+        } else {
+            // Check if all forced withdrawals are done
+            for (uint i = 0; i < tokens.length; i++) {
+                ready = ready && !exchange.isForcedWithdrawalPending(accountID, tokens[i].addr);
+            }
+        }
+
+        // Check that nothing is withdrawable anymore.
+        for (uint i = 0; i < tokens.length; i++) {
+            ready = ready && (exchange.getAmountWithdrawable(address(this), tokens[i].addr) == 0);
+        }
+
+        require(ready, "FUNDS_STILL_IN_EXCHANGE");
+
+        // TODO: withdraw proportional to the pool amount owned
+    }
+
+    struct AuxiliaryData
+    {
+        uint numItems;
+    }
+
+    // Processes work in the queue. Can only be called by the exchange owner
+    // before the blocks containing work for this pool are submitted.
+    // This just verifies if the work is done correctly and only then approves
+    // the L2 transactions that were already included in the block.
+    // Uses synchronized logic on L1/L2 to make the onchain logic easy and efficient.
+    function onSubmitBlocks(
+        ExchangeData.Block[] memory blocks,
+        uint                        blockIdx,
+        uint                        txIdx,
+        bytes                memory auxiliaryData
+        )
+        public
+        override
+        online
+        onlyExchangeOwner
+        returns (uint numTransactionsConsumed)
+    {
+        AuxiliaryData memory auxData = abi.decode(auxiliaryData, (AuxiliaryData));
+        ExchangeData.Block memory _block = blocks[blockIdx];
```

Comment:
Is it true that all the layer-2 join/exit transactions for an AMM account must be included into the same block?

Reply:
With the current implementation that's certainly the case. Makes things much easier and because the max amount of L2 transactions needed is just `2*numTokensInPool` regardless of the number of joins/exits it doesn't seem to be much of a limitation.